pkg/agent/controller/trafficcontrol/controller.go

```diff
@@ -0,0 +1,1039 @@
+}
+
+func (c *Controller) getTrafficControlPort(port *v1alpha2.TrafficControlPort) (uint32, bool) {
+	var portName string
+	switch {
+	case port.OVSInternal != nil:
+		portName = port.OVSInternal.Name
+	case port.Device != nil:
+		portName = port.Device.Name
+	case port.VXLAN != nil:
+		portName = genTunnelPortName(portNamePrefixVXLAN, port.VXLAN)
+	case port.GENEVE != nil:
+		portName = genTunnelPortName(portNamePrefixGENEVE, port.GENEVE)
+	case port.GRE != nil:
+		portName = genTunnelPortName(portNamePrefixGRE, port.GRE)
+	case port.ERSPAN != nil:
+		portName = genTunnelPortName(portNamePrefixERSPAN, port.ERSPAN)
+	}
+	if itf, ok := c.interfaceStore.GetInterfaceByName(portName); ok {
+		return uint32(itf.OFPort), true
+	}
+	return 0, false
+}
+
+func (c *Controller) deleteTrafficControlPort(port uint32) error {
+	if itf, ok := c.interfaceStore.GetInterfaceByOFPort(port); ok {
+		if err := c.ovsBridgeClient.DeletePort(itf.PortUUID); err != nil {
+			return err
+		}
+		c.interfaceStore.DeleteInterface(itf)
+	}
+	return nil
+}
+
+func (c *Controller) syncTrafficControl(tcName string) error {
+	startTime := time.Now()
+	defer func() {
+		klog.V(2).InfoS("Finished syncing TrafficControl", "TrafficControl", tcName, "durationTime", time.Since(startTime))
+	}()
+
+	tc, err := c.trafficControlLister.Get(tcName)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			// If a TrafficControl is removed and the corresponding state doesn't exist, just return.
+			tcState := c.getTCState(tcName)
+			if tcState == nil {
+				return nil
+			}
+			// If a TrafficControl is removed but the corresponding state exists, uninstall its dataplane configuration
+			// and remove the state.
+			if err = c.uninstallTrafficControl(tcName, tcState); err != nil {
+				return err
+			}
+			return nil
+		}
+		return err
+	}
+
+	tcState := c.getTCState(tcName)
+	// If corresponding state of a TrafficControl doesn't exist, create it.
+	if tcState == nil {
+		tcState = c.newTCState(tcName, tc.Spec.Action, tc.Spec.Direction)
+	}
+
+	var returnPort, targetPort uint32
+	var tcUpdate bool
+	var exists bool
+	c.ovsPortUpdateMutex.Lock()
+
+	if tc.Spec.ReturnPort != nil {
+		returnPort, exists = c.getTrafficControlPort(tc.Spec.ReturnPort)
+		// If the return port doesn't exist, create it.
+		if !exists {
+			if returnPort, err = c.createTrafficControlPort(tc.Spec.ReturnPort); err != nil {
+				return err
+			}
+		}
+
+		// Bind the created return port to the TrafficControl. If the return port is created firstly, install the OVS
+		// flow related to it.
+		if tcs, _ := c.installedTrafficControls.ByIndex(returnPortIndexName, strconv.Itoa(int(returnPort))); len(tcs) == 0 {
+			if err = c.ofClient.InstallTrafficControlReturnPortFlow(returnPort); err != nil {
+				return err
+			}
+		}
+
+		if tcState.returnPort != 0 && returnPort != tcState.returnPort {
+			notUsedByOthers, createdByTC := c.ifOFPortNotUsedByOthers(tcName, tcState.returnPort, returnPortIndexName)
```

Comment:
```suggestion
			shouldDeletePort, shouldDeleteFlow := c.shouldDeletePortAndFlow(tcName, tcState.returnPort, returnPortIndexName)
```
returning createdByTC as false is confusing when it's used by others but created by TC.

Reply:
Looks much better.